### DISCUSSION TITLE
Separate between single and multi replace?

### DISCUSSION THREAD

**Discussion by Nadyita** [spatie/regex#66] [Ideas]
I'm currently looking for a replacement for the built-in `preg_match()` and `preg_replace()` functions that is
- better with error-handling
- has better types

Your project ticks all the boxes except that `Regex::replace(…)->result()` returns `array | string` which triggers the same type-errors on phpstan and psalm as plain `preg_replace()` when simply assuming it must be a string, because the subject was a string.

My suggestion would be to distinguish between `replace` and `replaceMulti` and have them both return different `ReplaceResult`s, one having `->result()` return a string, the other an array of strings. Yes, this is mainly to make linters and autocompleters happy, but splitting a function that returns `array | string` into one that returns a string and one that returns an array makes it easier to work with.

If I implemented this and provided a PR for it, would you even consider it, or is this change too breaking for you or simply not fitting the goal?

**Comment by Nadyita** [comment]
Since no one has replied for a week, I will take this as a "not interested right now"

**Reply by pionl** [reply]
I think that discussions is not working on Github - I've posted some posts on multiple projects and no reaction :)
